add BeforeStep / AfterStep hooks

the3dsandwich opened the pull request:
As discussed in #280 

The `BeforeStep` and `AfterStep` hooks will enable code to be run before each `Given`, `When`, `Then`, etc. steps.

This way, we will be able to attach screenshots for every step.

vitalets commented on src/hooks/step.ts:
```diff
@@ -0,0 +1,153 @@
+}
+
+function getHookStepTitle(hook: GeneralStepHook) {
+  return hook.options.name ?? (hook.type === 'before' ? 'BeforeEach hook' : 'AfterEach hook');
```
```suggestion
  return hook.options.name ?? (hook.type === 'before' ? 'BeforeStep hook' : 'AfterStep hook');
```

the3dsandwich replied:
updated the name

vitalets commented on src/runtime/bddStepInvoker.ts:
```diff
@@ -46,16 +47,22 @@ export class BddStepInvoker {
-      return matchedDefinition.definition.fn.call(
+      const beforeHooksToRun = getStepHooksToRun('before', this.bddContext.tags);
+      await runStepHooks(beforeHooksToRun, { ...fixturesArg, $bddContext: this.bddContext });
+      const use = await matchedDefinition.definition.fn.call(
```
Let's name it `result` to not confuse with Playwright's `use()` function.

the3dsandwich replied:
changed the variable name to `result`

vitalets commented on src/runtime/bddStepInvoker.ts:
```diff
@@ -46,16 +47,22 @@ export class BddStepInvoker {
       // See: https://github.com/vitalets/playwright-bdd/issues/208
-      return matchedDefinition.definition.fn.call(
+      const beforeHooksToRun = getStepHooksToRun('before', this.bddContext.tags);
+      await runStepHooks(beforeHooksToRun, { ...fixturesArg, $bddContext: this.bddContext });
```
Here is a tricky place. If we make step hooks similar to scenario hooks, it means they can reference any custom fixture. For example:
```ts
BeforeStep(async ({ myCustomFixture }) => {
  // ...
}); 
```
This in turn means, that `myCustomFixture` should be instantiated in all tests and passed to all steps (in the spec generation phase):
```ts
// generated file
test('Check get started link', async ({ Given, When, Then, myCustomFixture }) => {
  await Given('I am on home page', null, { myCustomFixture });
  await When('I click link "Get started"', null, { myCustomFixture });
  await Then('I see in title "Installation"', null, { myCustomFixture });
});
```
This will make test gen process rather complex.
My suggestion for now is to start with the simplest approach: pass only `$testInfo` and `$step` fixtures to the step hooks. `$testInfo` will allow to attach screenshots.

The code could be changed as follows:
```ts
// currently pass only minimal set of fixtures to the step hooks: $testInfo, $test, $step, etc
const stepHooksFixturesArg = getBddAutoInjectFixtures(this.bddContext);
// ...
await runStepHooks(beforeHooksToRun, stepHooksFixturesArg);
// ...
await runStepHooks(afterHooksToRun, stepHooksFixturesArg);
```

We can enhance it in the future.

vitalets replied:
**UPDATE:**
I've realized, that my above solution is incorrect. We still need `page` fixture to make a screenshot.
Let's add tests for hook steps and then decide, how to manage fixtures.

the3dsandwich replied:
I noticed that in

```ts
await runStepHooks(afterHooksToRun, { ...fixturesArg, $bddContext: this.bddContext });
```

I need to add `...fixturesArg` to make `page` work

vitalets commented:
Let's add a test for step hooks in a new folder `test/hooks-step`. You can use `test/hooks-order-one-feature` as a reference. To run this particular test: `npm run only test/hooks-step`.

the3dsandwich commented:
> Let's add a test for step hooks in a new folder `test/hooks-step`. You can use `test/hooks-order-one-feature` as a reference. To run this particular test: `npm run only test/hooks-step`.

I added `test/hook-step`, renamed `use` to `result`, and renamed `getHookStepTitle` to `BeforeStep/AfterStep`.

regarding enabling page fixture in addition to common `this.bddContext`, any idea how we can proceed?

vitalets commented:
I've updated `test/hooks-step` a bit: removed scenario and worker hooks. Pushed the changes to your branch, so it's in the pr.
Now let me think a bit how to pass fixtures to the step hooks. Will post updates here.

the3dsandwich commented:
I found that I haven't changed `StepHookType` from `ScenarioHookType`. Renaming in this commit.

vitalets commented:
Hi @the3dsandwich 

I've pushed the fix for passing fixtures to step hooks, on top of your last commit.
Now we only need to write the docs about step hooks. 
- describe in the [Hooks](https://vitalets.github.io/playwright-bdd/#/writing-steps/hooks) section
- add to [API](https://vitalets.github.io/playwright-bdd/#/api) section

Are you willing to contribute to the docs as well?
If yes, you can run `npm run docs` and check edits locally.

the3dsandwich commented:
Working on it. Will push something ASAP

the3dsandwich commented:
I wasn't aware that tagging exist, so I'm unsure whether I got that part of the documentation right. Would you mind double-checking the contents are accurate?
In terms of docsify, did you install via `npm i -g docsify`?

vitalets commented on docs/writing-steps/hooks.md:
````diff
@@ -286,4 +288,114 @@ AfterScenario(async () => {
+const { BeforeStep } = createBdd();
+```
+
+Since Playwright-BDD **v8** you can target scenario hook to particular features/scenarios by `tags`:
````
```suggestion
You can target step hook to the steps of the specific feature/scenario by `tags`:
```

vitalets commented on .prototools:
```diff
@@ -0,0 +1,2 @@
+node = "~20"
```
What is this file for?

the3dsandwich replied:
My bad. I should have left this file out of the commit. It's to specify that this repo uses node 20 for this tool https://moonrepo.dev/proto

vitalets commented on docs/writing-steps/hooks.md:
```diff
@@ -286,4 +288,114 @@ AfterScenario(async () => {
+The hook function can accept **1 argument** - [test-scoped fixtures](https://playwright.dev/docs/test-fixtures#built-in-fixtures).
+You can access [$testInfo](https://playwright.dev/docs/api/class-testinfo), [$tags](writing-steps/bdd-fixtures.md#tags) and any built-in or custom fixtures. See more details in [BeforeScenario / Before API](api.md#beforescenario-before).
+
+#### Example of using `BeforeStep` with custom fixture
```
I suggest to put here exactly the use-case with screenshots:
```suggestion
#### Example of using `BeforeStep` to capture screenshot after each step
```

vitalets commented:
@the3dsandwich in your scenario with screenshots, if step fails do you expect the `AfterStep` still run?

the3dsandwich commented:
I personally do not expect `AfterStep` to run if a step fails and throws an error. If I want a screenshot of failed tests I would add it to the playwright config, or use a fixture to edit the title or something.

the3dsandwich commented:
I updated the docs a bit further, so that the example of taking a screenshot after every step is in the `AfterStep` block.

vitalets commented on docs/writing-steps/hooks.md:
````diff
@@ -398,4 +374,25 @@ AfterStep(async () => {
+export const { AfterStep } = createBdd(test);
+```
+
+Import `fixture.ts` in step definition
````
```suggestion
Import `fixtures.ts` in step definition
```

the3dsandwich replied:
fixed
